src/hooks/useResponsiveLayout.ts

```diff
@@ -38,16 +38,24 @@ export default function useResponsiveLayout(): ResponsiveLayoutResult {
     // This means it will only be defined if the component calling this hook is a child of a modal component. See BaseModal for the provider.
     const {activeModalType} = useContext(ModalContext);
 
-    // This refers to the state of the root navigator, and is true if and only if the topmost navigator is the "left modal navigator" or the "right modal navigator"
-    const isDisplayedInModalNavigator = !!useRootNavigationState(Navigation.isModalNavigatorActive);
+    // We are using these contexts directly instead of useNavigation/useNavigationState, because those will throw an error if used outside a navigator.
+    // This hook can be used within or outside a navigator, so using useNavigationState does not work.
+    // Furthermore, wrapping useNavigationState in a try/catch does not work either, because that breaks the rules of hooks.
+    // Note that these three lines are copied closely from the internal implementation of useNavigation: https://github.com/react-navigation/react-navigation/blob/52a3234b7aaf4d4fcc9c0155f44f3ea2233f0f40/packages/core/src/useNavigation.tsx#L18-L28
+    const navigationContainerRef = useContext(NavigationContainerRefContext);
+    const navigator = useContext(NavigationContext);
+    const currentNavigator = navigator ?? navigationContainerRef;
+
+    const isDisplayedInNarrowModalNavigator =
+        !!currentNavigator?.getParent?.(NAVIGATORS.RIGHT_MODAL_NAVIGATOR as unknown as undefined) || !!currentNavigator?.getParent?.(NAVIGATORS.LEFT_MODAL_NAVIGATOR as unknown as undefined);
```

Comment:
I kind of doubt that it's a significant performance problem because we know it's at most like 3 or 4 recursive calls, but I'll add a `useMemo` just to be safe.